File: packages/typia/src/transformers/features/misc/MiscLiteralsTransformer.ts
import ts from "typescript";

import { MiscLiteralsProgrammer } from "../../../programmers/misc/MiscLiteralsProgrammer";

import { ITransformProps } from "../../ITransformProps";
import { TransformerError } from "../../TransformerError";

export namespace MiscLiteralsTransformer {
  export const transform = (
    props: Omit<ITransformProps, "modulo">,
  ): ts.Expression => {
    // CHECK GENERIC ARGUMENT EXISTENCE
    if (!props.expression.typeArguments?.[0])
      throw new TransformerError({
        code: "typia.misc.literals",
        message: "generic argument is not specified.",
      });

    // GET TYPE INFO
    const node: ts.TypeNode = props.expression.typeArguments[0];
    const type: ts.Type = props.context.checker.getTypeFromTypeNode(node);

    if (type.isTypeParameter())
      throw new TransformerError({
        code: "typia.misc.literals",
        message: "non-specified generic argument.",
      });

    // DO TRANSFORM
    return MiscLiteralsProgrammer.write({
      context: props.context,
      type,
    });
  };
}

File: packages/typia/native/transform/features/misc/MiscLiteralsTransformer.go
package misc

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  nativeprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers/misc"
  nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type miscLiteralsTransformerNamespace struct{}

var MiscLiteralsTransformer = miscLiteralsTransformerNamespace{}

func (miscLiteralsTransformerNamespace) Transform(props nativetransform.ITransformProps) *shimast.Node {
  if props.Expression == nil || props.Expression.TypeArguments == nil || len(props.Expression.TypeArguments.Nodes) == 0 {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    "typia.misc.literals",
      Message: "generic argument is not specified.",
    }))
  }
  node := props.Expression.TypeArguments.Nodes[0]
  typ := props.Context.Checker.GetTypeFromTypeNode(node)
  if typ != nil && typ.IsTypeParameter() {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    "typia.misc.literals",
      Message: "non-specified generic argument.",
    }))
  }
  return nativeprogrammers.MiscLiteralsProgrammer.Write(nativeprogrammers.MiscLiteralsProgrammer_IProps{
    Context: props.Context,
    Type:    typ,
  })
}
